sim/state.ts

```diff
@@ -0,0 +1,340 @@
+	serializeBattle(battle: Battle): /* Battle */ AnyObject {
+		const state: /* Battle */ AnyObject = this.serialize(battle, BATTLE, battle);
+		state.field = this.serializeField(battle.field);
+		state.sides = new Array(battle.sides.length);
+		for (const [i, side] of battle.sides.entries()) {
+			state.sides[i] = this.serializeSide(side);
+		}
+		state.prng = battle.prng.seed;
+		state.hints = Array.from(battle.hints);
+		return state;
+	}
+
+	// Deserialization can only really be done on the root Battle object as
+	// the leaf nodes like Side or Pokemon contain backreferences to Battle
+	// but don't contain the information to fill it in because the cycles in
+	// the graph have been serialized as references. Once deserialzized, the
+	// Battle can than be restarted (and provided with a `send` function for
+	// receiving updates).
+	deserializeBattle(serialized: string | /* Battle */ AnyObject): Battle {
+		const state: /* Battle */ AnyObject =
+			typeof serialized === 'string' ? JSON.parse(serialized) : serialized;
+		const options = {
+			formatid: state.format,
+			seed: state.prngSeed,
+			rated: state.rated,
+			debug: state.debugMode,
+			// We need to tell that Battle we're creating that it's been
+			// deserialized so that it allows us to populate it correctly and
+			// doesn't attempt to start playing out until we're ready.
+			deserialized: true,
+			strictChoices: state.strictChoices,
+		};
+		for (const side of state.sides) {
+			const team = side.team.split(side.team.length > 9 ? ',' : '');
+			// @ts-ignore - index signature
+			options[side.id] = {
+				name: side.name,
+				avatar: side.avatar,
+				team: team.map((p: string) => side.pokemon[Number(p) - 1].set),
+			};
+		}
+		// We create the Battle, allowing it to instantiate the Field/Side/Pokemon
+		// objects for us. The objects it creates will be incorrect, but we descend
+		// down through the fields and repopulate all of the objects with the
+		// correct state.
+		const battle = new Battle(options);
+		this.deserialize(state, battle, BATTLE, battle);
+		this.deserializeField(state.field, battle.field);
+		for (const [i, side] of state.sides.entries()) {
+			this.deserializeSide(side, battle.sides[i]);
+		}
+		battle.prng = new PRNG(state.prng);
+		// @ts-ignore - readonly
+		battle.hints = new Set(battle.hints);
+		return battle;
+	}
+
+	serializeField(field: Field): /* Field */ AnyObject {
+		return this.serialize(field, FIELD, field.battle);
+	}
+
+	private deserializeField(state: /* Field */ AnyObject, field: Field) {
+		this.deserialize(state, field, FIELD, field.battle);
+	}
+
+	serializeSide(side: Side): /* Side */ AnyObject {
+		const state: /* Side */ AnyObject = this.serialize(side, SIDE, side.battle);
+		state.pokemon = new Array(side.pokemon.length);
+		const team = new Array(side.pokemon.length);
+		for (const [i, pokemon] of side.pokemon.entries()) {
+			state.pokemon[i] = this.serializePokemon(pokemon);
+			team[side.team.indexOf(pokemon.set)] = i + 1;
```

Comment:
Any reason to use one-based indices here? ``side.team`` isn't used for anything after the battle is started, so it probably doesn't deserve this treatment.

Reply:
Because I find it more readable and it looks more like the `team` command you send the simulator.

Reply:
But why would you want to read ``state.team`` ? According to this implementation, it wouldn't match the actual ``/team`` command sent, wouldn't it?

Reply:
I believe that meaning is not valuable enough for humans... for us to care for its readability... But I don't think this is worth arguing over either.

Reply:
If you think *this* isn't worth it, wait until you see what I've been working on this morning! While I'm debugging this, I constantly need to read over a `control.json` and `test.json`, and `activeRequest` on `sides` is a big block of annoying redundant info to scroll over, so now I'm working to collapse *that*. That's significantly more work than team encoding, but is of similar impact on readability. Readability matters, because I literally can't make progress on this if I can't debug WTF is going on, and I can only debug it if the output is readable.


(I'm actually not sure about how much work it will be to collapse `activeRequest` now that we mutate it thanks to removing `|callback|`. I think it would be ideal if the respective `getRequestData` will return the updated request automatically, but if not my plan is to elide `activeRequest` and instead encode an `activeRequestDiff` which will usually be `{}`, but in the cases where `activeRequest` was updated it will contain information about what things were revealed to be `disabled`/`trapped` etc.).

Reply:
> Readability matters, because I literally can't make progress on this if I can't debug WTF is going on

Since you say it like so, this is literally like adding ``console.log()`` statements all over the place. The final protocol should be clean and needn't have the debug features.